modules/core/src/system.cpp

```diff
@@ -1300,6 +1300,407 @@ void setUseCollection(bool flag)
+
+        if(pStruct->enableMapping)
+        {   // Critical section
+            cv::AutoLock guard(pStruct->mutex); // Guard from concurrent child creation
+            pChild = pTLS->pCurrentNode->findChild(payload);
+            if(!pChild)
+            {
+                pChild = new Node<NodeData>(payload);
+
+                if(pTLS->pCurrentNode->m_payload.m_mtRegion)
+                {
+                    pChild->m_payload.m_mtRegion = true;
+                    pChild->m_payload.m_pTls = new cv::TLSData<NodeDataTLS>;
+                }
+
+                if(m_mtPoint)
+                    pChild->m_payload.m_mtRegion = true;
+
+                pTLS->pCurrentNode->addChild(pChild);
+            }
+        }
+        else
+        {
+            pChild = pTLS->pCurrentNode->findChild(payload);
+            if(!pChild)
+            {
+                pTLS->pCurrentNode->m_payload.m_stopPoint = true;
+                pStruct->pRefNode = pTLS->pCurrentNode; // set ref node for thraded nodes to stop correctly
+                return;
+            }
+        }
+        pTLS->pCurrentNode = pChild;
+        pTLS->pCurrentNode->m_payload.m_counter++;
+
+        // Start threads junction
+        if(m_mtPoint)
+        {
+            CV_Assert(pStruct->pRefNode == NULL);
+            pStruct->pRefNode = pTLS->pCurrentNode;  // set current reference node
+        }
+
+        m_regionTicks = getTickCount();
+    }
+}
+
+// Recursevly process nodes and collect TLS data across threads
+static void consolidateThreadsData(cv::instr::InstrNode *pNode)
+{
+    if(pNode->m_payload.m_pTls)
+    {
+        std::vector<cv::instr::NodeDataTLS*> treadsVector;
+        pNode->m_payload.m_pTls->gather(treadsVector);
+
+        unsigned long long ticksMax = 0;
+        for(size_t i = 0; i < treadsVector.size(); i++)
+        {
+            ticksMax = (ticksMax < treadsVector[i]->m_ticksAcc)?treadsVector[i]->m_ticksAcc:ticksMax;
+            treadsVector[i]->m_ticksAcc = 0; // reset thread data
+        }
+        pNode->m_payload.m_ticksAcc += ticksMax;
+    }
+
+    for(size_t i = 0; i < pNode->m_childs.size(); i++)
+        consolidateThreadsData(pNode->m_childs[i]);
+}
+
+static void resetNodePointer()
+{
+    InstrStruct *pStruct = &getInstrumentStruct();
+
+    // Reset node pointer for threads
+    std::vector<cv::instr::InstrTLSStruct*> treadsVector;
+    pStruct->tlsStruct.gather(treadsVector);
+    for(size_t i = 0; i < treadsVector.size(); i++)
+        treadsVector[i]->pCurrentNode = NULL;
+
+    // Restore node pointer for main thread
+    pStruct->tlsStruct.get()->pCurrentNode = pStruct->pRefNode;
+    pStruct->pRefNode = NULL;
+}
+
+IntrumentationRegion::~IntrumentationRegion()
+{
+    InstrStruct *pStruct = &getInstrumentStruct();
+    if(pStruct->useInstr)
+    {
+        if(!m_disabled)
+        {
+            if(std::uncaught_exception())
+            {
+                // Detect bad exit from IW
+            }
```

Comment:
I had a thought to detect bad exits from some functions and mark it in the trace, but it was never implemented.